### PR TITLE
Bugfix/#544 missing files on page

### LINKED ISSUE #544: Uploaded files are not listed in the upload field when editing

### PR DESCRIPTION
**Pull request by todd-dembrey**
Pretty substantial rework of how the multi-file upload works. Allows users to upload new files and clear existing files in the same action.

closes #544

### CODE REVIEW

**Comment by todd-dembrey** [comment]
image

**Comment by frjo** [comment]
Mostly works but therein one bug when you have multiple files uploaded and tries to delete one of them. The wrong file gets deleted.

It looks like the get numbered in different order.

**Comment by todd-dembrey** [comment]
Interesting, It seemed to work fine in my local testing. Will try again and see if i can get it to break.

**Comment by todd-dembrey** [comment]
Looks like I was relying on the data being ordered in the dict from the form. This is not correct. have a fix.

**Comment by todd-dembrey** [comment]
@frjo Thanks for the review. Improvements made and tests added.

File: opentech/apply/stream_forms/fields.py [hunk 1]
```diff
@@ -1,25 +1,64 @@
...
+            upload = files.getlist(name)
+        else:
+            upload = files.get(name)
+            if not isinstance(upload, list):
+                upload = [upload]
+
+        checkbox_name = self.clear_checkbox_name(name)
+        checkboxes = {k for k in data if checkbox_name in k}
+        cleared = {
+            i for i, checkbox in enumerate(checkboxes)
```
Comment (frjo) [inline opentech/apply/stream_forms/fields.py]: "i" will always start at 0 regardless of the count of the checkbox checked.

What ever checkbox you check the first firewall get deleted. If you check any two checkboxes the first two files will get deleted.

Comment (todd-dembrey) [reply opentech/apply/stream_forms/fields.py]: You've rebased over my fixes... Will force push over your changes

**Comment by todd-dembrey** [comment]
@frjo Rebased all the changes onto master including the fixes you have rebased away.

**Comment by frjo** [comment]
Did I manage to forget to pull! Sorry about that. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by frjo** [comment]
Tested and now it works.

**Pull request merged**